### ISSUE TITLE
Security vs Usability 

### ISSUE THREAD

**Issue by creslinux** [freqtrade/freqtrade#847]
For discussion. 
The Freqtrade project is opensource and has an active community and all code is peer reviewed. 
These are good and proper barriers to protect a users data. I have personally reviewed the code end to end and found no malicious content.

In common with any and all automated trading programs the user must give a level of trust to the program to not abuse their API keys, which must be protected with as many layers the user and project can make available. 

However unlikely it is possible for rogue code, core project accounts to become compromised/utilised to inject functionality to scrape users exchange api keys. 

In the event of API key compromise API accounts on exchanges may commonly be restricted to a subset of features so direct withdrawal of funds may be prevented. This would not prevent all capital being used to purchase/pump a low cap coin. To the user the impact may be just a devastating.

A good measure to prevent API key leaking to protect the project from itself is for the user to be able to tightly control egress data from the bot. Ideally this should make use of code and control from outside the project. The rational being it is exponentially difficult for a threat agent to compromise many projects in parallel. 

It is common in Freqtrade for deployment to be within a docker container. 

I would propose discovery work be undertaken to offer the user a tested and documented deployment model/environment that wrappers network control with software that is outside the projects control, which the project will non-functionally test/approve to provide a functioning install. 

A contender for a deployment blueprint to prevent egress of data to any target other than an exchange *may be:

1 - Freqtrade installed into named Freqtrade docker instance as normal 
2 - 2nd squidguard docker installed from widely deployed maintained repository along with dns cache
2a - simple squidguard configuration to only allow HTTPS and  Only to a whitelist of Exchanges 
2b - simple DNS configuration to only allow DNS resolution of Exchange whitelist 
3 - host OS firewall script to forward all data from Docker Freqtrade to Docker Squidguard. This is simple in Linux. 

2a and 2b must be and can be simple and open for the user to maintain themselves. 

Whilst this solution is off-the-shelf and a proven topology to prevent data leaking through DNS calls or directly sent another site not whitelisted there would be useability and functional compromise. 

Not least the compromise to usability includes telegram integration would cease. 
There is no simple means to allow telegram integration and also tightly control egress data flows. 

If the project would aim to support an increased secure deployment environment as an option to users another means to control the bot should be provided / made available.

**Comment by xmatthias** [comment]
I'm a big security-fan myself - but i think the above is slightly excessive and focuses on the wrong things:

there are 2 things that will ALWAYS be necessary to run the bot:

* you will always need to allow https calls to the exchange you use
* you will always need some way to access the server (probably ssh) to maintain / start / stop / update the bot.

In theory, you can control your outbound traffic as much as you'd like - if someone compromises the server in some other way, you're screwed, as that attacker can then use the (allowed) traffic to the remote exchange to trade all your coins away.

Technically, there are many possibilities (be that squidguard, systemd (it can limit network calls to a specific IP only)).
However the best technical solution will fail if the server itself is not properly secured - and someone can run commands on your server.

I personally see it out of scope for this project to suggest how to secure a server - however we could add some links in the documentation to projects focused on exactly this.

## alternative idea

Not sure if some exchanges allow it - but maybe you can limit an API key to a specific IP (assuming your server IP will never change). This (in combination with a properly secured server) will provide the best security you can have - as your key can only be used from your server (which only you can access - as it's properly secured). 
With this in place, you would not have to worry about egress data.

**Comment by creslinux** [comment]
I completely agree with the two connections that will always be needed - outbound to exchange and inbound to control. 

A challenge I can't easily square is that the main means to control the bot is telegram, by opening egress to telegram it restricts the validity of any outbound whitelist/control to prevent leaking data. Whilst a touch point of scope is how a Freqtrade is wrappered a requirement that means it cannot easily be meaningfully achieved could be counter productive. 

I feel another local control panel/means mechanism not telegram should be provided. 
Perhaps it already exists and I've missed it

Agree there are many attack vectors. 
This one is to provide a means for the user to protect themselves from the project.  A poorly hardened OS or other vulnerable attack surface would be out of scope, for the provider of that software to recommend as they tend to do.

**Comment by jblestang** [comment]
What about making telegram Optional as a separate process that would only have access to the SQLite file as a starter?

**Comment by xmatthias** [comment]
The whole RPC logic is implemented to allow different methods (other than telegram) of communication.
only Telegram is implemented at the moment - but feel free to implement another method if you see the need.

However that adding a rest-service (or whatever else) will again open a channel into the bot - so you would only shift the attack vector from telegram to something else.

@jblestang that would kill all control possibility - only allowing for reporting (`/profit`,  `/profit table` , ...)- but removing `/start`, `/stop`,  `/forcesell`, which are commands (as these need to start/stop/sell within the bot).

**Comment by creslinux** [comment]
I'll do some reading, see if I can craft a local loopback RPC client. 
Could be useful to have in general outside of the security idea if clean

Thanks for the feedback!

**Comment by xmatthias** [comment]
have a look at freqtrade/rpc/* on how it's implemented ;)

a command line client would be nice (then you can leave telegram switched off if that's a security concern ...)

**Comment by creslinux** [comment]
Sounds like a perfect learning project - im up for that!
